### DISCUSSION TITLE
utility.batchAll should't fire a lot of utility.ItemCompleted

### DISCUSSION THREAD

**Discussion by nazar-pc** [paritytech/substrate#10290] [Ideas]
Conceptually `utility.batchAll` is a wrapper that makes sure one of two is true: all transactions succeeded or all rolled back.

With that in mind, having `utility.ItemCompleted` event for every single transaction in a batch seems wasteful.

Thoughts?

**Comment by xlc** [comment]
It is required to correctly associate events with calls.

**Reply by nazar-pc** [reply]
I understand that, but as a developer you know that you have sent a batch, so while `utility.ItemCompleted` is useful for `utility.batch`, in case of `utility.batchAll` you know you have sent a batch and only need to care whether it succeeded as a whole, you don't necessarily need to know about it as many times as number of extrinsics in a batch, it is just wasteful.

**Reply by xlc** [reply]
No. I mean the completed event is required to identify which sub call trigger which event.

**Reply by nazar-pc** [reply]
Well, many extrinsics will have their own events anyway, so I don't really understand why it is necessary.

**Reply by xlc** [reply]
Take this example:

```
batch([
	batch([ transfer(Alice, 100) ]),
	batch([ transfer(Alice, 100) ])
])
```

Now there is only 1 transfer event. Can you tell which transfer emit the event? (without knowing the implementation details of transfer)

Note that batch will always success even the inner transfer failed.

You should read the issue / PR that add the item completed event for more context.

**Reply by nazar-pc** [reply]
I'm talking specifically about `batchAll`, for which `utility.BatchCompleted` must be sufficient. I'm not sure why you keep bringing `batch` here. I did check https://github.com/paritytech/substrate/pull/9462 and understand how it it useful for `batch` call, but `batchAll` by definition either succeeds in full or doesn't go through at all.

**Reply by xlc** [reply]
```
batchAll([
	batch([ transfer(Alice, 100) ]),
	batch([ transfer(Alice, 100) ])
])
```
Now there is only 1 transfer event. Can you tell which transfer emit the event? (without knowing the implementation details of transfer)

**Reply by nazar-pc** [reply]
There are still events from `batch` call, so you can check if there is `BatchInterrupted` and if so which batch it corresponds to.

Since `batch` succeeds either way, I don't see how the fact that `batchAll` emits `ItemCompleted` is helpful here at all. We essentially have 4 `ItemCompleted` here instead of 2 and that serves literally no purpose unless I'm missing something obvious.
